### PR TITLE
Generate configuration cache key based on the tooling API request

### PR DESCRIPTION
**Pull request by adammurdoch**
### Context

- For task execution, use the requested tasks to generate the cache key. This is the same key as if the tasks were run from the command line.
- For model requests, use the model name plus requested tasks, if any, to generate the cache key.
- For build action requests, use a dummy key plus the requested tasks, if any. Currently, the client action is not exposed by the tooling API client and so its details cannot be used to generate a cache key (at least, not without some nasty work arounds). A later PR can improve this calculation.

For all of the above, include the init scripts and other inputs as were previously included for task execution.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

File: subprojects/launcher/src/main/java/org/gradle/launcher/exec/RunActionRequirements.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+public class RunActionRequirements extends AbstractToolingModelRequirements {
+    public RunActionRequirements(StartParameterInternal startParameter,
+                                 boolean runsTasks) {
+        super(startParameter, runsTasks);
+    }
+
+    @Override
+    public void appendKeyTo(Hasher hasher) {
+        // Identify the type of action
+        hasher.putByte((byte) 3);
```
Comment (lptr) [inline subprojects/launcher/src/main/java/org/gradle/launcher/exec/RunActionRequirements.java:31]: We have `Hasher.signature(Class)` to create a `HashCode` for the type, and then you can use `hasher.putHash()`. It is a little bit more costly, but it avoids a whole suite of problems with colliding IDs.

We could also move this thing to the execution engine, and mark these things as `@Input` properties, and let it handle caching on its own.

**Pull request merged**